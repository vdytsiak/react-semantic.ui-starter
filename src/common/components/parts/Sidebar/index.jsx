// @flow
import React, {Component} from 'react'
import {Menu, Icon} from 'semantic-ui-react'
import {NavLink} from 'react-router-dom'
import {
	StyledSidebar,
	SidebarLogo,
	SidebarLogoContainer,
	SidebarItem,
	SidebarLogoutItem
} from './style'
import {Spacer} from 'styles/base'
import type {RouteItem} from 'types'

type Props = {
	open: boolean,
	logout: () => void,
	routing: Array<RouteItem>,
	isMobile: boolean
}

class SidebarComponent extends Component {
	props: Props

	render () {
		const {open, logout, routing, isMobile} = this.props

		const sidebarProps = {
			visible: open || !isMobile,
			as: Menu,
			vertical: true,
			icon: 'labeled',
			animation: 'push',
			width: 'thin'
		}

		const routes = routing.map((route, i) => {
			const {external, path, icon, name, strict, exact} = route
			// Props that are common for both "<a>" and "RR Link"
			const basicProps = {
				as: external ? 'a' : NavLink,
				link: true,
				[external ? 'href' : 'to']: path
			}

			// Is it's RR Link, then it needs additional props
			const externalProps = external
				? {}
				: {
					strict,
					exact,
					activeClassName: 'active'
				}

			// Summarize
			const propsMenuItem = {
				...externalProps,
				...basicProps
			}

			return (
				<SidebarItem key={i} {...propsMenuItem} icon>
					<Icon name={icon} /> {name}
				</SidebarItem>
			)
		})

		return (
			<StyledSidebar {...sidebarProps}>
				<SidebarLogoContainer href="https://github.com/Metnew/suicrux">
					<SidebarLogo alt="logo" shape="circular" centered />
				</SidebarLogoContainer>
				{routes}
				<Spacer />
				<SidebarLogoutItem onClick={logout}>
					<Icon name="sign out" />
					Logout
				</SidebarLogoutItem>
			</StyledSidebar>
		)
	}
}

export default SidebarComponent
