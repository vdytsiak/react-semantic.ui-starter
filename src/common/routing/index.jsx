export history from './history'
export {getSidebarRoutes, getRouterRoutes, getRoutes} from './routes'
