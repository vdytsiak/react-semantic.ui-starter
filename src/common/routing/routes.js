// @flow
import React from 'react'
import {Route, Redirect} from 'react-router-dom'
import RouteAuth from 'components/addons/RouteAuth'
import {asyncComponent} from 'react-async-component'
import {Loader, Dimmer, Header, Icon} from 'semantic-ui-react'
import _ from 'lodash'
import type {RouteItem} from 'types'

function asyncComponentCreator (url) {
	return asyncComponent({
		// flow-disable-next-line: The parameter passed to import() must be a literal string
		resolve: () => import(/* webpackChunkName:"[index].[request]" */ `containers/${url}/index.jsx`),
		LoadingComponent () {
			return (
				<Dimmer active>
					<Loader size="large" active>
						Loading page...
					</Loader>
				</Dimmer>
			)
		},
		ErrorComponent () {
			return (
				<Dimmer active>
					<Header inverted as="h2" icon textAlign="center">
						<Icon name="refresh" />
						Refresh
						<Header.Subheader>Got error while loading page.</Header.Subheader>
					</Header>
				</Dimmer>
			)
		},
		autoResolveES2015Default: true,
		env: process.env.BROWSER ? 'browser' : 'node',
		serverMode: 'resolve'
	})
}

function routingFnCreator (useFor: 'sidebar' | 'routing' | 'all' = 'all') {
	const [AsyncDashoard, AsyncLinks, AsyncLogin, AsyncNotFound] = [
		'Dashboard',
		'Links',
		'Login',
		'NotFound'
	].map(a => {
		return asyncComponentCreator(a)
	})

	const routes: Array<RouteItem> = [
		{
			path: '/',
			exact: true,
			icon: 'newspaper',
			name: 'Dashboard',
			sidebarVisible: true,
			tag: RouteAuth,
			component: AsyncDashoard
		},
		{
			path: '/links',
			name: 'Links',
			exact: true,
			icon: 'bookmark',
			sidebarVisible: true,
			tag: RouteAuth,
			component: AsyncLinks
		},
		{
			external: true,
			path: 'https://github.com/Metnew/react-semantic.ui-starter',
			icon: 'github',
			name: 'Github',
			sidebarVisible: true
		},
		{
			path: '/auth',
			name: 'Auth',
			exact: true,
			tag: Route,
			component: AsyncLogin
		},
		// Find the way to add/remove routes conditionally
		{
			name: '404',
			tag: RouteAuth,
			component: AsyncNotFound
		},
		{
			tag: Redirect,
			to: '/auth'
		}
	]

	const fns = {
		// Returns routing for sidebar menu
		sidebar (x: Array<RouteItem> = routes) {
			return x
				.filter(a => a.sidebarVisible)
				.map(a =>
					_.pick(a, ['path', 'name', 'icon', 'external', 'strict', 'exact'])
				)
		},
		// Returns routing for React-Router
		routing (x: Array<RouteItem> = routes) {
			return x
				.filter(a => !!a.tag)
				.map(a =>
					_.pick(a, [
						'path',
						'name',
						'strict',
						'exact',
						'component',
						'tag',
						'to'
					])
				)
		},
		all () {
			return routes
		}
	}

	return fns[useFor]
}

export const getRoutes = routingFnCreator()
export const getSidebarRoutes = routingFnCreator('sidebar')
export const getRouterRoutes = routingFnCreator('routing')
