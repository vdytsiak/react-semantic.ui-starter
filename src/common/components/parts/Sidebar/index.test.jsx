import React from 'react'
import Sidebar from 'components/parts/Sidebar'
import {shallow} from 'enzyme'
import toJson from 'enzyme-to-json'
import {getSidebarRoutes} from 'routing'

const props = {
	isMobile: true,
	logout: () => ({}),
	routing: getSidebarRoutes(),
	open: true
}

describe('Sidebar component', () => {
	it('same snapshot for Sidebar', () => {
		const component = shallow(<Sidebar {...props} />)
		expect(toJson(component)).toMatchSnapshot()
	})
})
