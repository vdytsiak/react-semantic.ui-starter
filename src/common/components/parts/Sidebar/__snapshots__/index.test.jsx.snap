// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component same snapshot for Sidebar 1`] = `
<style__StyledSidebar
  animation="push"
  as={[Function]}
  icon="labeled"
  vertical={true}
  visible={true}
  width="thin"
>
  <style__SidebarLogoContainer
    href="https://github.com/Metnew/suicrux"
  >
    <style__SidebarLogo
      alt="logo"
      centered={true}
      shape="circular"
    />
  </style__SidebarLogoContainer>
  <style__SidebarItem
    activeClassName="active"
    as={[Function]}
    exact={true}
    icon={true}
    key="0"
    link={true}
    to="/"
  >
    <Icon
      as="i"
      name="newspaper"
    />
     
    Dashboard
  </style__SidebarItem>
  <style__SidebarItem
    activeClassName="active"
    as={[Function]}
    exact={true}
    icon={true}
    key="1"
    link={true}
    to="/links"
  >
    <Icon
      as="i"
      name="bookmark"
    />
     
    Links
  </style__SidebarItem>
  <style__SidebarItem
    as="a"
    href="https://github.com/Metnew/react-semantic.ui-starter"
    icon={true}
    key="2"
    link={true}
  >
    <Icon
      as="i"
      name="github"
    />
     
    Github
  </style__SidebarItem>
  <base__Spacer />
  <style__SidebarLogoutItem
    onClick={[Function]}
  >
    <Icon
      as="i"
      name="sign out"
    />
    Logout
  </style__SidebarLogoutItem>
</style__StyledSidebar>
`;
